File: agent/integrations/e2b.py
"""E2B sandbox backend implementation."""

from __future__ import annotations

import logging
import os
from typing import Any

from deepagents.backends.protocol import (
    ExecuteResponse,
    FileDownloadResponse,
    FileUploadResponse,
    SandboxBackendProtocol,
    WriteResult,
)
from deepagents.backends.sandbox import BaseSandbox

logger = logging.getLogger(__name__)

DEFAULT_E2B_TEMPLATE = "opencode"
DEFAULT_E2B_TIMEOUT = 300
ANTHROPIC_API_KEY_ENV = "ANTHROPIC_API_KEY"


def _combine_output(stdout: Any, stderr: Any) -> str:
    parts = [str(part) for part in (stdout, stderr) if part]
    return "\n".join(parts)


def _build_sandbox_envs() -> dict[str, str]:
    api_key = os.environ.get(ANTHROPIC_API_KEY_ENV)
    if not api_key:
        return {}
    return {ANTHROPIC_API_KEY_ENV: api_key}


def _resolve_timeout(timeout: int | None) -> int:
    if timeout is not None:
        return timeout
    env_timeout = os.environ.get("SANDBOX_TIMEOUT")
    if env_timeout:
        try:
            return int(env_timeout)
        except ValueError:
            logger.warning(
                "Invalid SANDBOX_TIMEOUT=%r, falling back to %s",
                env_timeout,
                DEFAULT_E2B_TIMEOUT,
            )
    return DEFAULT_E2B_TIMEOUT


class E2BBackend(BaseSandbox):
    """E2B backend implementation conforming to SandboxBackendProtocol."""

    def __init__(self, sandbox: Any) -> None:
        self._sandbox = sandbox
        self._default_timeout = DEFAULT_E2B_TIMEOUT

    @property
    def id(self) -> str:
        sandbox_id = getattr(self._sandbox, "sandbox_id", None)
        if not sandbox_id:
            sandbox_id = getattr(self._sandbox, "id", None)
        if not sandbox_id:
            msg = "E2B sandbox did not expose an id"
            raise AttributeError(msg)
        return sandbox_id

    def execute(self, command: str, *, timeout: int | None = None) -> ExecuteResponse:
        effective_timeout = timeout if timeout is not None else self._default_timeout
        try:
            result = self._sandbox.commands.run(command, timeout=effective_timeout)
        except Exception as exc:
            exit_code = getattr(exc, "exit_code", None)
            if exit_code is None:
                raise
            output = _combine_output(getattr(exc, "stdout", ""), getattr(exc, "stderr", ""))
            if not output:
                output = str(exc)
            return ExecuteResponse(output=output, exit_code=exit_code, truncated=False)

        output = _combine_output(getattr(result, "stdout", ""), getattr(result, "stderr", ""))
        return ExecuteResponse(
            output=output,
            exit_code=getattr(result, "exit_code", 0),
            truncated=False,
        )

    def write(self, file_path: str, content: str) -> WriteResult:
        try:
            self._sandbox.files.write(file_path, content)
            return WriteResult(path=file_path, files_update=None)
        except Exception as exc:
            return WriteResult(error=f"Failed to write file '{file_path}': {exc}")

    def download_files(self, paths: list[str]) -> list[FileDownloadResponse]:
        responses: list[FileDownloadResponse] = []
        for path in paths:
            try:
                content = self._sandbox.files.read(path)
                responses.append(FileDownloadResponse(path=path, content=content, error=None))
            except Exception as exc:
                responses.append(FileDownloadResponse(path=path, content="", error=str(exc)))
        return responses

    def upload_files(self, files: list[tuple[str, bytes]]) -> list[FileUploadResponse]:
        responses: list[FileUploadResponse] = []
        for path, content in files:
            try:
                self._sandbox.files.write(path, content)
                responses.append(FileUploadResponse(path=path, error=None))
            except Exception as exc:
                responses.append(FileUploadResponse(path=path, error=str(exc)))
        return responses


def _update_thread_sandbox_metadata(sandbox_id: str) -> None:
    """Update thread metadata with sandbox_id."""
    try:
        import asyncio

        from langgraph.config import get_config
        from langgraph_sdk import get_client

        config = get_config()
        thread_id = config.get("configurable", {}).get("thread_id")
        if not thread_id:
            return
        client = get_client()

        async def _update() -> None:
            await client.threads.update(
                thread_id=thread_id,
                metadata={"sandbox_id": sandbox_id},
            )

        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            asyncio.run(_update())
        else:
            loop.create_task(_update())
    except Exception:
        pass


def create_e2b_sandbox(
    sandbox_id: str | None = None,
    timeout: int | None = None,
) -> SandboxBackendProtocol:
    """Create or connect to an E2B sandbox."""
    try:
        from e2b import Sandbox
    except ImportError as exc:
        raise ImportError("e2b is not installed. Please run `uv sync --all-extras`.") from exc

    template = os.environ.get("E2B_TEMPLATE", DEFAULT_E2B_TEMPLATE)
    envs = _build_sandbox_envs()
    resolved_timeout = _resolve_timeout(timeout)
    api_key = os.environ.get("E2B_API_KEY")

    if sandbox_id:
        sandbox = Sandbox.connect(sandbox_id)
        if hasattr(sandbox, "set_timeout"):
            sandbox.set_timeout(resolved_timeout)
    else:
        kwargs: dict[str, Any] = {"template": template, "timeout": resolved_timeout}
        if api_key:
            kwargs["api_key"] = api_key
        if envs:
            kwargs["envs"] = envs
        sandbox = Sandbox.create(**kwargs)

    _update_thread_sandbox_metadata(sandbox.sandbox_id)
    return E2BBackend(sandbox)

File: src/integrations/e2b.ts
/**
 * E2B sandbox integration.
 * Mirrors agent/integrations/e2b.py
 *
 * E2B is a cloud sandbox provider for running code:
 * https://e2b.dev
 *
 * Env vars:
 *   E2B_API_KEY       — API key for E2B (optional, defaults to SDK default)
 *   E2B_TEMPLATE      — sandbox template (default: "opencode")
 *   SANDBOX_TIMEOUT   — default timeout in seconds (default: 300)
 *   ANTHROPIC_API_KEY — forwarded into sandbox as env var if set
 */

import type {
  ExecuteResponse,
  FileDownloadResponse,
  FileInfo,
  FileOperationError,
  FileUploadResponse,
  GrepMatch,
  SandboxBackendProtocol,
  WriteResult,
} from "deepagents";

const DEFAULT_E2B_TEMPLATE = "opencode";
const DEFAULT_E2B_TIMEOUT = 300;

// ─── Helpers ──────────────────────────────────────────────────────────────────

function resolveTimeout(timeout?: number): number {
  if (timeout !== undefined) return timeout;
  const env = process.env.SANDBOX_TIMEOUT;
  if (env) {
    const n = Number.parseInt(env, 10);
    if (!Number.isNaN(n)) return n;
  }
  return DEFAULT_E2B_TIMEOUT;
}

function buildSandboxEnvs(): Record<string, string> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  return apiKey ? { ANTHROPIC_API_KEY: apiKey } : {};
}

function shellQuote(s: string): string {
  return `'${s.replace(/'/g, "'\\''")}'`;
}

function combineOutput(stdout: string, stderr: string): string {
  return [stdout, stderr].filter(Boolean).join("\n");
}

// ─── E2BBackend ───────────────────────────────────────────────────────────────

export class E2BBackend implements SandboxBackendProtocol {
  // biome-ignore lint/suspicious/noExplicitAny: E2B SDK types not available at compile time
  private readonly _sandbox: any;
  private readonly _defaultTimeout: number;

  // biome-ignore lint/suspicious/noExplicitAny: E2B SDK types
  constructor(sandbox: any, defaultTimeout: number = DEFAULT_E2B_TIMEOUT) {
    this._sandbox = sandbox;
    this._defaultTimeout = defaultTimeout;
  }

  get id(): string {
    const sid = this._sandbox?.sandboxId ?? this._sandbox?.id;
    if (!sid) throw new Error("E2B sandbox did not expose an id");
    return String(sid);
  }

  // ─── Execute ───────────────────────────────────────────────────────────────

  async execute(command: string): Promise<ExecuteResponse> {
    try {
      const result = await this._sandbox.commands.run(command, {
        timeout: this._defaultTimeout,
      });
      const output = combineOutput(result?.stdout ?? "", result?.stderr ?? "");
      return { output, exitCode: result?.exitCode ?? 0, truncated: false };
    } catch (err: unknown) {
      const exitCode = (err as Record<string, unknown>)?.exitCode as
        | number
        | undefined;
      if (exitCode !== undefined) {
        const stdout =
          ((err as Record<string, unknown>)?.stdout as string) ?? "";
        const stderr =
          ((err as Record<string, unknown>)?.stderr as string) ?? "";
        const output = combineOutput(stdout, stderr) || String(err);
        return { output, exitCode, truncated: false };
      }
      return {
        output: err instanceof Error ? err.message : String(err),
        exitCode: 1,
        truncated: false,
      };
    }
  }

  // ─── File operations ───────────────────────────────────────────────────────

  async read(filePath: string, offset = 0, limit = 500): Promise<string> {
    try {
      const result = await this.execute(
        `sed -n '${offset + 1},${offset + limit}p' ${shellQuote(filePath)}`,
      );
      return result.output;
    } catch (err) {
      return `Error reading file: ${err instanceof Error ? err.message : String(err)}`;
    }
  }

  // biome-ignore lint/suspicious/noExplicitAny: generic FileData
  async readRaw(filePath: string): Promise<any> {
    const text = await this.read(filePath);
    return { content: new TextEncoder().encode(text), mime_type: "text/plain" };
  }

  async write(filePath: string, content: string): Promise<WriteResult> {
    try {
      await this._sandbox.files.write(filePath, content);
      return { path: filePath, error: undefined };
    } catch (err) {
      return {
        path: filePath,
        error: `Failed to write file '${filePath}': ${err instanceof Error ? err.message : String(err)}`,
      };
    }
  }

  async edit(
    filePath: string,
    oldString: string,
    newString: string,
    replaceAll = false,
    // biome-ignore lint/suspicious/noExplicitAny: EditResult from deepagents
  ): Promise<any> {
    const content = await this.read(filePath);
    const occurrences = content.split(oldString).length - 1;
    if (occurrences === 0) {
      return {
        error: `String not found in ${filePath}`,
        path: filePath,
        occurrences: 0,
      };
    }
    const updated = replaceAll
      ? content.split(oldString).join(newString)
      : content.replace(oldString, newString);
    const writeResult = await this.write(filePath, updated);
    return {
      error: writeResult.error,
      path: filePath,
      occurrences: replaceAll ? occurrences : 1,
    };
  }

  async lsInfo(path: string): Promise<FileInfo[]> {
    const result = await this.execute(
      `ls -la ${shellQuote(path)} 2>&1 | tail -n +2`,
    );
    if (result.exitCode !== 0) return [];
    return result.output
      .split("\n")
      .filter(Boolean)
      .map((line) => {
        const parts = line.split(/\s+/);
        const name = parts[parts.length - 1] ?? "";
        const isDir = line.startsWith("d");
        return {
          path: `${path.replace(/\/$/, "")}/${name}${isDir ? "/" : ""}`,
          size: Number.parseInt(parts[4] ?? "0", 10) || 0,
          is_dir: isDir,
        } satisfies FileInfo;
      });
  }

  async grepRaw(
    pattern: string,
    path?: string | null,
    glob?: string | null,
  ): Promise<GrepMatch[] | string> {
    const pathArg = path ? shellQuote(path) : ".";
    const globArg = glob ? `--include=${shellQuote(glob)}` : "";
    const result = await this.execute(
      `grep -rn ${globArg} ${shellQuote(pattern)} ${pathArg} 2>/dev/null`,
    );
    if (result.exitCode !== 0) return [];
    return result.output
      .split("\n")
      .filter(Boolean)
      .map((line) => {
        const [file, lineNum, ...rest] = line.split(":");
        return {
          path: file ?? "",
          line: Number.parseInt(lineNum ?? "0", 10),
          text: rest.join(":"),
        } satisfies GrepMatch;
      });
  }

  async globInfo(pattern: string, path = "/"): Promise<FileInfo[]> {
    const result = await this.execute(
      `find ${shellQuote(path)} -path ${shellQuote(pattern)} 2>/dev/null`,
    );
    if (result.exitCode !== 0) return [];
    return result.output
      .split("\n")
      .filter(Boolean)
      .map((p) => ({
        path: p,
        size: 0,
        is_dir: false,
        last_modified: "",
        mime_type: "application/octet-stream",
      }));
  }

  async downloadFiles(paths: string[]): Promise<FileDownloadResponse[]> {
    return Promise.all(
      paths.map(async (p) => {
        try {
          const content = await this._sandbox.files.read(p);
          return {
            path: p,
            content,
            error: null,
          } satisfies FileDownloadResponse;
        } catch {
          return {
            path: p,
            content: null,
            error: "file_not_found" as FileOperationError,
          } satisfies FileDownloadResponse;
        }
      }),
    );
  }

  async uploadFiles(
    files: Array<[string, Uint8Array]>,
  ): Promise<FileUploadResponse[]> {
    return Promise.all(
      files.map(async ([path, data]) => {
        try {
          const content = new TextDecoder().decode(data);
          await this._sandbox.files.write(path, content);
          return { path, error: null } satisfies FileUploadResponse;
        } catch (_err) {
          return {
            path,
            error: "permission_denied" as FileOperationError,
          } satisfies FileUploadResponse;
        }
      }),
    );
  }
}

// ─── Factory ──────────────────────────────────────────────────────────────────

/**
 * Create or reconnect to an E2B sandbox.
 */
export async function createE2bSandbox(
  sandboxId?: string,
  timeout?: number,
): Promise<SandboxBackendProtocol> {
  // biome-ignore lint/suspicious/noExplicitAny: dynamic import from optional peer dep
  let Sandbox: any;

  try {
    const sdk = await import("e2b" as never);
    Sandbox = sdk.Sandbox;
  } catch {
    throw new Error(
      "E2B requires the 'e2b' npm package. Install it with: bun add e2b",
    );
  }

  const template = process.env.E2B_TEMPLATE ?? DEFAULT_E2B_TEMPLATE;
  const resolvedTimeout = resolveTimeout(timeout);
  const envs = buildSandboxEnvs();
  const apiKey = process.env.E2B_API_KEY;

  // biome-ignore lint/suspicious/noExplicitAny: dynamic SDK
  let rawSandbox: any;

  if (sandboxId) {
    rawSandbox = await Sandbox.connect(sandboxId);
    if (typeof rawSandbox.setTimeout === "function") {
      rawSandbox.setTimeout(resolvedTimeout);
    }
  } else {
    const opts: Record<string, unknown> = {
      template,
      timeout: resolvedTimeout,
    };
    if (apiKey) opts.apiKey = apiKey;
    if (Object.keys(envs).length) opts.envs = envs;
    rawSandbox = await Sandbox.create(opts);
  }

  return new E2BBackend(rawSandbox, resolvedTimeout);
}
